larpandoracontent/LArMonitoring/EventClusterValidationAlgorithm.cc

```diff
@@ -151,181 +249,681 @@ void EventClusterValidationAlgorithm::GetHitParents(
         }
     }
 
+    if (m_handleDeltaRays)
+    {
+        for (auto &[pCaloHit, parents] : hitParents)
+        {
+            parents.m_pMainMC = this->FoldPotentialDeltaRayTo(pCaloHit, parents.m_pMainMC);
+        }
+    }
+
     for (const Cluster *const pCluster : clusters)
     {
         const CaloHitList &isolatedHits{pCluster->GetIsolatedCaloHitList()};
         CaloHitList clusterCaloHits;
         pCluster->GetOrderedCaloHitList().FillCaloHitList(clusterCaloHits);
         clusterCaloHits.insert(clusterCaloHits.end(), isolatedHits.begin(), isolatedHits.end());
+
         for (const CaloHit *const pCaloHit : clusterCaloHits)
         {
-            // Ignoring the calo hit if truth matching is missing
             if (hitParents.find(pCaloHit) == hitParents.end())
+            {
                 continue;
+            }
             hitParents[pCaloHit].m_pCluster = pCluster;
         }
     }
 }
 
 //------------------------------------------------------------------------------------------------------------------------------------------
 
+const MCParticle* EventClusterValidationAlgorithm::FoldMCTo(const MCParticle *const pMC) const
+{
+    if (!this->IsEM(pMC))
+    {
+        return pMC;
+    }
+
+    bool hasAncestorElectron{false};
+    const MCParticle *pCurrentMC{pMC};
+    const MCParticle *pLeadingMC{pMC};
+    while (!pCurrentMC->IsRootParticle())
+    {
+        const MCParticle *const pParentMC{*(pCurrentMC->GetParentList().begin())};
+        const int parentPdg{std::abs(pParentMC->GetParticleId())};
+        if (parentPdg == PHOTON || parentPdg == E_MINUS)
+        {
+            if (parentPdg == E_MINUS)
+            {
+                hasAncestorElectron = true;
+            }
+            pCurrentMC = pParentMC;
+            continue;
+        }
+        pLeadingMC = pCurrentMC;
+        break;
+    }
+
+    // Don't fold "showers" that consist of only compton scatters
+    // Trying to prevents distant diffuse hits disconnected from a "real" bremm + pair production shower being clustered together
+    if (hasAncestorElectron || std::abs(pLeadingMC->GetParticleId()) == E_MINUS)
+    {
+        return pLeadingMC;
+    }
+    else if (this->CausesShower(pLeadingMC, 0))
+    {
+        return pLeadingMC;
+    }
+    else
+    {
+        return pMC;
+    }
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+bool EventClusterValidationAlgorithm::CausesShower(const MCParticle *const pMC, int nDescendentElectrons) const
```

Comment:
Does this behave logically if showers are folded at sim-level? Do we care?

Reply:
It doesnt matter what this does for already folded MC as the two options for folding, pMC and pLeadingMC, will just be the same particle. Ideally you wouldnt configure folding if you know your sample is already folded but I think it behaves normally (does nothing) even if you do.